Move to CherryPy

ashcrow opened the pull request:
- No more greenlets
- No more gevent
- CherryPy now used as the application server
- Moves calls from directly using the ```etcd.Client``` to utilizing the CherryPy plugin bus

mbarnes commented on requirements.txt:
```diff
@@ -4,4 +4,4 @@ requests
 python-etcd==0.4.2
 setuptools
-ansible>=2.0.0.2
+ansible>=2.0.1.0
```
This is fairly close to bleeding edge.  Probably should sync the ansible requirement in `commissaire.spec`.

ashcrow replied:
Good point. However versions before 2.0.1.0 have a breaking bug :cry:.

ashcrow replied:
I went ahead and bumped the spec up as well. Looks like 2.0.1.0 is being built in EPEL already.

mbarnes commented on src/commissaire/cherrypy_plugins.py:
```diff
@@ -0,0 +1,52 @@
+            store = etcd.Client(**self.store_kwargs)
+            return (False, store.set(key, json_entity))
+        except:
+            return (True, [])
```
Dunno if this return type is following some convention, but I wonder if returning the exception object itself would be generally more useful to callers.  (I come from the GObject world, which has a similar pub/sub pattern.)  Something like:
```
    try:
        ...
        return (store.set(key, json_entity), None)
    except Exception as ex:
        return ([], ex)
```

ashcrow replied:
No specific convention. I copied it from Go more or less. I'm happy with making a change of returning the exception. :+1:.

mbarnes commented on src/commissaire/cherrypy_plugins.py:
```diff
@@ -0,0 +1,52 @@
+
+    def store_save(self, key, json_entity):
+        try:
+            store = etcd.Client(**self.store_kwargs)
```
I assume you create the `etcd.Client` instance here instead of in `__init__` because it's failable and might throw an exception, right?  Do we really need to create a new instance for each call though?  Seems expensive.

What about:
```
    def __init__(self, bus, store_kwargs):
        plugins.SimplePlugin.__init__(self, bus)
        self.store_kwargs = store_kwargs
        self.store = None

    def _get_store(self):
        if not self.store:
            self.store = etcd.Client(**self.store_kwargs)
        return self.store

    ...
```

ashcrow replied:
I don't think we need to, no. I like this idea. :+1:

mbarnes commented on src/commissaire/handlers/clusters.py:
```diff
@@ -145,6 +150,8 @@ def on_put(self, req, resp, name):
             cluster = Cluster(status='ok', hostset=[])
             etcd_resp = self.store.set(key, cluster.to_json(secure=True))
+            # _, etcd_resp = cherrypy.engine.publish(
+            #    'store-save', key, cluster.to_json(secure=True))[0]
```
I'm looking at http://docs.cherrypy.org/en/latest/extend.html#bus-api and the docs don't indicate the return type of `publish`.  What else is it returning that we need the `[0]`?

ashcrow replied:
It **always** returns a list. So if ```publish``` has a ```return True``` the caller will get ```[True]```. If you return ```True, True``` the caller will get ```[(True, True)]```.

mbarnes replied:
Huh.  Weird.

mbarnes commented on src/commissaire/script.py:
```diff
@@ -254,32 +234,45 @@ def main():  # pragma: no cover
         target=investigator, args=(INVESTIGATE_QUEUE, config, store_kwargs))
     PROCS['investigator'].start()
+
+    # Add app config to the config to the global config for use elsewhere
```
nit: Wording

ashcrow replied:
:laughing: :+1:

ashcrow commented:
:arrow_up:

mbarnes commented on src/commissaire/script.py:
```diff
@@ -235,7 +235,8 @@ def main():  # pragma: no cover
 
-    # Add app config to the config to the global config for use elsewhere
+    # Add our config instance to the cherrypy global config so we can use it's
+    # values elsewhere
```
I feel bad for even mentioning this, but `it's` -> `its`

mbarnes commented:
Looks good.  Did you have more changes pending for the commented-out parts or is this merge-ready?

ashcrow commented:
I have a few more changes I'd like to make before we merge. Should be ready to go Monday or Tuesday morning.

ashcrow commented:
@mbarnes fixup is ready for review. This moves everything **but** delete to the plugin system. Assuming everything looks good here I'll follow up with a delete.

mbarnes commented:
Doesn't have to be done now, but all these messy looking publish calls are screaming out to me for some simpler wrapper functions, maybe in `handlers/util.py` or even establish a place for global utilities.

Something like:
```
def get_etcd_key(key):
    return cherrypy.engine.publish('store-get', key)[0]

def set_etcd_key(key, value, **kwargs):
    return cherrypy.engine.publish('store-save', key, value, **kwargs)[0]
```
Perhaps also have the functions re-raise the `Exception` object if they get one and just return the `EtcdResult`, so callers can still use `try ... except` blocks as normal.  I *think* that would look like:
```
def get_etcd_key(key):
    resp, exc = cherrypy.engine.publish('store-get', key)[0]
    if exc:
        raise exc
    return resp

def set_etcd_key(key, value, **kwargs):
    resp, exc = cherrypy.engine.publish('store-save', key, value, **kwargs)[0]
    if exc:
        raise exc
    return resp
```

mbarnes commented:
Apart from that the code changes look good.  I admit I skimmed through the test code changes.  The unit tests pass but I am getting a repeatable E2E test failure:
```
 Scenario: Retrieve status with authentication  # features/status/retrieve.feature:8
    Given we have a valid username and password  # features/steps/__init__.py:23 0.000s
    When we get status                           # features/steps/status.py:6 0.238s
    Then commissaire will allow access           # features/steps/__init__.py:30 0.000s
    And commissaire will return status           # features/steps/status.py:12 0.000s
      Traceback (most recent call last):
        File "/home/mbarnes/commissaire/lib/python2.7/site-packages/behave/model.py", line 1456, in run
          match.run(runner.context)
        File "/home/mbarnes/commissaire/lib/python2.7/site-packages/behave/model.py", line 1903, in run
          self.func(context, *args, **kwargs)
        File "features/steps/status.py", line 17, in impl
          assert key in results.keys()
      AssertionError
      
      Captured stdout:
      [u'etcd', u'investigator']
      
      Captured logging:
      INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 127.0.0.1



Failing scenarios:
  features/status/retrieve.feature:8  Retrieve status with authentication
```

ashcrow commented:
@mbarnes interesting. I'll take a look tomorrow morning at the e2e test failure.

ashcrow commented:
Reproduced the e2e failure. Will fix.

ashcrow commented:
@mbarnes I've fixed the e2e test. I've found two more and am looking at them now. Will note when ready to check again.

ashcrow commented:
FWIW here is an example of one failing.

```
  Given we have a valid username and password                   # features/steps/__init__.py:23 0.000s
    And we have a cluster named honeynut                          # features/steps/cluster.py:11 0.272s
    And we have a host at 10.2.0.2                                # features/steps/hosts.py:21t per request.
    And we have a host at 10.2.0.2                                # features/steps/hosts.py:21 0.270sating.
    And we have added host 10.2.0.2 to cluster honeynut           # features/steps/cluster.py:19 0.295s
    And we have deleted host 10.2.0.2                             # features/steps/hosts.py:32
    And the provided cluster total hosts is 0                     # features/steps/cluster.py:120 0.000sh 10.2.0.2...
    And the provided cluster total hosts is 0                     # features/steps/cluster.py:120ster honeynut
    Given we have a valid username and password                   # features/steps/__init__.py:23 0.000s
    And we have added host 10.2.0.2 to cluster honeynut           # features/steps/cluster.py:19 0.296s
    And we have deleted host 10.2.0.2                             # features/steps/hosts.py:3219per request.
    And the provided data is []                                   # features/steps/__init__.py:68 0.000sh 10.2.0.2...
    And the provided data is []                                   # features/steps/__init__.py:68ster honeynut
    Given we have a valid username and password                   # features/steps/__init__.py:23 0.000s
    And we have added host 10.2.0.2 to cluster honeynut           # features/steps/cluster.py:19 0.295s
    And we have deleted host 10.2.0.2                             # features/steps/hosts.py:3219per request.
    And commissaire will note it's not found                      # features/steps/__init__.py:38 0.000sh 10.2.0.2...
    And commissaire will note it's not found                      # features/steps/__init__.py:38ster honeynut
    And commissaire will note a conflict                      # features/steps/__init__.py:38 0.000s
    And commissaire will note a conflict                      # features/steps/__init__.py:38
    Given we have a valid username and password                # features/steps/__init__.py:23 0.000s
    And the provided cluster total hosts is 1                  # features/steps/cluster.py:120 0.000s
2016-03-23 13:55:13,143 - transport - 3452 -  139796405176064 - WARNING - UNREACHABLE 10.2.0.2: setupequest.
2016-03-23 13:55:13,144 - investigator - 3452 -  139796405176064 - WARNING - Getting info failed for 10.2.0.2: Can not run for 10.2.0.2
2016-03-23 13:55:13,148 - investigator - 3452 -  139796405176064 - INFO - 10.2.0.2 is now in investigating.
2016-03-23 13:55:13,149 - investigator - 3452 -  139796405176064 - INFO - Wrote key for 10.2.0.2
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 258, in _bootstrap

    Given we have a valid username and password                # features/steps/__init__.py:23 0.000s
    And we have a cluster named honeynut                       # features/steps/cluster.py:113
    self.run()
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "commissaire/src/commissaire/jobs/investigator.py", line 87, in investigator
    data = json.loads(etcd_resp.value)
    And the provided data is ['10.2.0.2']                      # features/steps/__init__.py:68 0.000s
Exception EtcdWatchTimedOut: Watch timed out: ReadTimeoutError("HTTPConnectionPool(host='127.0.0.1', port=54014): Read timed out.",)
[...]
etcd.EtcdWatchTimedOut: Watch timed out: ReadTimeoutError("HTTPConnectionPool(host='127.0.0.1', port=54014): Read timed out.",)
```

I have a feeling it will end up being how we are watching the investigator upon multiple runs (since the calls are queued) ... but we'll find out soon enough :smile:.

ashcrow commented:
@mbarnes :arrow_up:

mbarnes commented on src/commissaire/jobs/clusterexec.py:
```diff
@@ -71,12 +63,15 @@ def clusterexec(cluster_name, command):
-    etcd_resp = store.get('/commissaire/clusters/{0}'.format(cluster_name))
+    etcd_resp, _ = cherrypy.engine.publish(
+        'store-get', '/commissaire/clusters/{0}'.format(cluster_name))[0]
+    print etcd_resp.value
```
This is `print` intentional or leftover debugging?

ashcrow replied:
Left over from debugging. I believe I removed it in the next fixup.

mbarnes commented on test/test_jobs_clusterexec.py:
```diff
@@ -51,35 +51,42 @@ def test_clusterexec(self):
+                    [[return_value, None]],
+                    [[return_value, None]],
+                    [[return_value, None]],
+                )
```
Passing thought: Going forward, I wonder if we should try to annotate these side effects somehow so they more easily map to points in the source code.  These things always take me the longest to figure out when debugging a failed test or adding new ones.  It's almost guesswork at times.

ashcrow replied:
Good idea. To be honest, that is what I was ending up having so much trouble with on the port. The actual functionality was pretty straight forward.

mbarnes commented:
Looks good, and I verified the tests all pass now.

@cgwalters-bot +r
@ashcrow  Do you mind if I play around with those wrapper functions I mentioned, or would you prefer not?

ashcrow commented:
@mbarnes Let's do the wrapper functions when we get back to reworking the util module.

I think you got the r and + backwards there :smile:

mbarnes commented:
@cgwalters-bot r+
